plugins/modules/source_control/gitlab/gitlab_project_variable.py

```diff
@@ -162,106 +246,150 @@ def list_all_project_variables(self):
+            "masked": var_obj.get('masked'), "protected": var_obj.get('protected'),
+            "variable_type": var_obj.get('variable_type')
         }
-        if environment_scope is not None:
-            var["environment_scope"] = environment_scope
-        return self.project.variables.create(var)
 
-    def update_variable(self, key, var, value, masked, protected, variable_type, environment_scope):
-        if (var.value == value and var.protected == protected and var.masked == masked
-                and var.variable_type == variable_type
-                and (var.environment_scope == environment_scope or environment_scope is None)):
-            return False
+        if var_obj.get('environment_scope') is not None:
+            var["environment_scope"] = var_obj.get('environment_scope')
+
+        self.project.variables.create(var)
+        return True
 
+    def update_variable(self, var_obj):
         if self._module.check_mode:
             return True
+        self.delete_variable(var_obj)
+        self.create_variable(var_obj)
+        return True
 
-        if (var.protected == protected and var.masked == masked
-                and var.variable_type == variable_type
-                and (var.environment_scope == environment_scope or environment_scope is None)):
-            var.value = value
-            var.save()
+    def delete_variable(self, var_obj):
+        if self._module.check_mode:
             return True
-
-        self.delete_variable(key)
-        self.create_variable(key, value, masked, protected, variable_type, environment_scope)
+        self.project.variables.delete(var_obj.get('key'), filter={'environment_scope': var_obj.get('environment_scope')})
         return True
 
-    def delete_variable(self, key):
-        if self._module.check_mode:
-            return
-        return self.project.variables.delete(key)
+
+def compare(requested_variables, existing_variables, state):
+    # we need to do this, because it was determined in a previous version - more or less buggy
+    # basically it is not necessary and might results in more/other bugs!
+    # but it is required  and only relevant for check mode!!
+    # logic represents state 'present' when not purge. all other can be derived from that
+    # untouched => equal in both
+    # updated => name and scope are equal
+    # added => name and scope does not exist
+    untouched = list()
+    updated = list()
+    added = list()
+
+    if state == 'present':
+        existing_key_scope_vars = list()
+        for item in existing_variables:
+            existing_key_scope_vars.append({'key': item.get('key'), 'environment_scope': item.get('environment_scope')})
+
+        for var in requested_variables:
+            if var in existing_variables:
+                untouched.append(var)
+            else:
+                compare_item = {'key': var.get('name'), 'environment_scope': var.get('environment_scope', '*')}
```

Comment:
If in the above code the default value `'*'` must not be there, why must it be here?
```suggestion
                compare_item = {'key': var.get('name'), 'environment_scope': var.get('environment_scope')}
```

Reply:
Ah yes. I think I implement it like this because of [this](https://github.com/ansible-collections/community.general/pull/4038#discussion_r793308583).

I re-run the integration test and see if all sill pass.

Reply:
Unfortunately GitHub doesn't want to show me what "this" is :)